### PR TITLE
gossiper: do_status_check: allow evicting dead nodes from membership with no host_id

### LINKED ISSUE #16936: Gossip round FAIL: std::runtime_error (Host 127.0.23.4 does not have HOST_ID application_state)
Seen in https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7321/artifact/logs-full.release.000/1705976646543_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_simple_add_new_node_while_schema_changes_with_stream/node1.log:

```
INFO  2024-01-23 02:14:19,915 [shard 0:strm] gossip - failure_detector_loop: Mark node 127.0.23.4 as DOWN
INFO  2024-01-23 02:14:19,915 [shard 0:strm] gossip - InetAddress 127.0.23.4 is now DOWN, status = BOOT
INFO  2024-01-23 02:14:27,913 [shard 0: gms] gossip - FatClient 127.0.23.4 has been silent for 30000ms, removing from gossip
INFO  2024-01-23 02:14:27,915 [shard 0: gms] gossip - Removed endpoint 127.0.23.4
WARN  2024-01-23 02:14:27,916 [shard 0: gms] gossip - === Gossip round FAIL: std::runtime_error (Host 127.0.23.4 does not have HOST_ID application_state)
```

\cc @bhalevy

### PR DESCRIPTION
**Pull request by bhalevy**
The short series allows do_status_check to handle down nodes that don't have HOST_ID application state.

Fixes #16936

### CODE REVIEW

**Comment by bhalevy** [comment]
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7462/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split006___test_add_node_with_large_partition4/ looks like just flakiness.

Looking also into the unit test failures...

**Comment by bhalevy** [comment]
@kbr-scylla please review

File: gms/gossiper.cc [hunk 1]
```diff
@@ -757,11 +757,12 @@ future<> gossiper::do_status_check() {
 
         // check for dead state removal
         auto expire_time = get_expire_time_for_endpoint(endpoint);
-        const auto host_id = get_host_id(endpoint);
-        if (!is_alive && (now > expire_time)
-             && (!get_token_metadata_ptr()->is_normal_token_owner(host_id))) {
+        if (!is_alive && (now > expire_time)) {
+          const auto host_id = get_host_id(*eps);
+          if (!host_id || !get_token_metadata_ptr()->is_normal_token_owner(host_id)) {
```
Comment (kbr-scylla) [inline gms/gossiper.cc:762]: I'm worried that https://github.com/scylladb/scylladb/issues/16936 indicates a real bug that we have, and this change would hide it.

The bug would be that an endpoint appears in gossiper state which does not have a Host ID.
My question is how can this happen?

I thought it's an invariant we preserve that every endpoint comes with a Host ID.
Whenever node starts gossiping, it always inserts its host ID.
And also after a full cluster restart, when nodes are gossiping their peers' state restored from disk, they also include their host IDs.

So isn't there some larger problem we're dealing with? That host IDs are disappearing?

Comment (bhalevy) [reply gms/gossiper.cc:762]: We can add an internal error in non-release modes to help diagnose the problem in testing and follow-through, as suggested here, in release mode so not to interfere with production.

**Comment by bhalevy** [comment]
In v2 ([4f87401](https://github.com/scylladb/scylladb/commit/4f874014fc1993a411c95fe869d8f65fa35649f8)):
- rebased
- gms: endpoint_state: add get_host_id
- gossiper: do_status_check: allow evicting dead nodes from membership
  - use eps->get_host_id
  - call on_internal_error_noexcept if !host_id to abort in testing and follow through in production

**Comment by bhalevy** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7984/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/)
> 
> #### Failed Tests (4/25267):
> * [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
> * [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)

https://github.com/scylladb/scylladb/issues/17028

> * [test_cassandra_issue_10250](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/%28root%29/test_concurrent_schema/Tests___Unit_Tests___test_cassandra_issue_10250) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cassandra_issue_10250)
> * [topology.test_concurrent_schema.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_concurrent_schema_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_concurrent_schema.debug.2)
> 
> #### Build Details:
> * Duration: 2 hr 8 min
> * Builder: spider1.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/14612#issuecomment-1632291426 / https://github.com/scylladb/scylladb/issues/14312

**Review by kbr-scylla — APPROVED** [review]

**Comment by kbr-scylla** [comment]
LGTM, but waiting until `next` is unfreezed.

**Comment by kbr-scylla** [comment]
> The short series allows do_status_check to handle down nodes that hove HOST_ID application state.

s/hove/don't have/?

Also, could you please rebase the series so we could rerun CI on latest master with dtest-raft-topology too?

**Comment by bhalevy** [comment]
- fixed typo in PR description
- rebased
- gossiper: get_host_id: differentiate between no endpoint_state and no application_state
- gossiper: print the host_id when endpoint state goes UP/DOWN

**Comment by bhalevy** [comment]
@kbr-scylla I'm not sure about backport.
Please advise

**Review by kbr-scylla — APPROVED** [review]

**Comment by kbr-scylla** [comment]
I see no evidence that backport is necessary.

**Pull request merged**